### PR TITLE
Update Android java-part

### PR DESCRIPTION
**Pull request by MoNTE48**
**Issues:** https://github.com/minetest/minetest/issues/7813
https://github.com/minetest/minetest/issues/7529
https://github.com/minetest/minetest/issues/7501
https://github.com/minetest/minetest/issues/2736

**What changed:**
reformat code with Android Studio standart,
updated target sdk,
updated gradlew,
fixed display aspect on 18+/:9 dispalys (like a Samsung Galaxy S9),
removed small app icons, they are not required,
fixed xml in unpacking activity,
minor fixes in manifest.

**What new:**
Support android permission: on Android 6.0+ you need to manually give write permission (_as required by google_),
Background during unpacking (_just a demo, I advise you to choose a beautiful image, such as a block of grass or dirt_),
Material Design. There is no more terrible Android 2 interface,
Immersive mode (_Android 4.4+ - hide NavBar for fullscreen mode_),

Maybe something else...
_(I did not make any changes to MinetestAssetCopy.java and MinetestTextEntry.java. Minor change for full screen mode is not considered :)
Strings were moved automatically according to Android Studio standards. This does not affect the work.)_

### CODE REVIEW

**Comment by stujones11** [comment]
@MoNTE48 Thank you very much for sharing this, I will test this as soon as possible tomorrow.

**Comment by juozaspo** [comment]
@MoNTE48 I made a build from latest commit on this PR, works fine on my Huawei Honor 7 Lite (NEM-L21), doesnt exit to phone interface when entering & leaving lock screeen.
screenshot_20181029-072944

**Comment by stujones11** [comment]
Tested on SG Tab-A (Android 7) and Honor 9 lite (Android 8)

Installs and runs okay on both devices, except I still seeing many error messages like this in the logs
```
E/MinetestAssetCopy: Failed to open asset file "/storage/emulated/0/Minetest/builtin/common/filterlist.lua" for size check
E/MinetestAssetCopy: Failed to open asset file "/storage/emulated/0/Minetest/builtin/common/vector.lua" for size check
E/MinetestAssetCopy: Failed to open asset file "/storage/emulated/0/Minetest/builtin/common/strict.lua" for size check
E/MinetestAssetCopy: Failed to open asset file "/storage/emulated/0/Minetest/builtin/common/serialize.lua" for size check
```

Also, the virtual keyboard is causing the same crash on Android 8 as I reported here https://github.com/minetest/minetest/issues/7813#issuecomment-433646090 Your suggestion did work for me but does not appear to be included in this PR.

One other observation is that this is still creating a Minetest directory in the storage root which not good practice these days. As far as I know, we should be using `net.minetest.minetest` in the user data directory. This was everything is properly cleaned up when a user uninstalls the app.

I think that 5.0.0 gives us the opportunity to correct this while also solving the problem of the game and builtin scripts not updating properly when upgrading from 0.4.17. Being able to migrate existing single-player worlds would be a nice option to have though I don't think there will be much requirement. I think the majority of mobile players connect to remote servers rather than single-player mode.

**Comment by MoNTE48** [comment]
1) **"E/MinetestAssetCopy: Failed to open asset file"** -need a few tests. Can you confirm that this does not happen on the current build of Minetest 5.0.0? (Not for 0.4.17 on Play Store, 5.0.0!) API 26-28 may introduce too many undocumented changes.
The best option is to use the MultiCraft implementation. We do not compare files. We do unpacking when changing the build number. This speeds up the launch and we had no problems like the one above. Although you should know that this implementation is not perfect ;)
2) **Virtual Keyboard** My trick was supposed to work. The input field should be on a transparent background. This was incompatible with the setting of the orientation of the activity in the Manifest, so I removed it and made an orientation change in Java. For me, it worked. It is very strange that you still have a crash. I will check it.

**Comment by stujones11** [comment]
@MoNTE48 Indeed the Asset file size check error is present is all existing builds, I thought this PR would have fixed it, however, it does not really cause any problems I am aware of.

With the virtual keyboard, the crash report may be slightly different, any clues?
```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: net.minetest.minetest, PID: 18401
    java.lang.RuntimeException: Unable to start activity ComponentInfo{net.minetest.minetest/net.minetest.minetest.MinetestTextEntry}: java.lang.IllegalStateException: Only fullscreen activities can request orientation
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3194)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3302)
        at android.app.ActivityThread.-wrap12(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1891)
        at android.os.Handler.dispatchMessage(Handler.java:108)
        at android.os.Looper.loop(Looper.java:166)
        at android.app.ActivityThread.main(ActivityThread.java:7425)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:245)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:921)
     Caused by: java.lang.IllegalStateException: Only fullscreen activities can request orientation
        at android.os.Parcel.readException(Parcel.java:1954)
        at android.os.Parcel.readException(Parcel.java:1892)
        at android.app.IActivityManager$Stub$Proxy.setRequestedOrientation(IActivityManager.java:5683)
        at android.app.Activity.setRequestedOrientation(Activity.java:6121)
        at net.minetest.minetest.MinetestTextEntry.onCreate(MinetestTextEntry.java:24)
        at android.app.Activity.performCreate(Activity.java:7372)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1218)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3147)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3302) 
        at android.app.ActivityThread.-wrap12(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1891) 
        at android.os.Handler.dispatchMessage(Handler.java:108) 
        at android.os.Looper.loop(Looper.java:166) 
        at android.app.ActivityThread.main(ActivityThread.java:7425) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:245) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:921) 
```

I am not  really sure about the lock screen issue, perhaps @juozaspo can answer this?

**Comment by MoNTE48** [comment]
@stujones11 
1) check the Manifest, is it exactly the same as here? https://github.com/minetest/minetest/pull/7820/files#diff-63ebd89fd84762840e8d7598266394ffR50
2) remove this line https://github.com/minetest/minetest/pull/7820/files#diff-3292ececbf56785f308dadd30769da09R24

**Comment by stujones11** [comment]
@MoNTE48 

1. Yes.
2. Works for me.

> Google does a lot of weird things.

Yes, thank you for the explanation.

**Comment by MoNTE48** [comment]
Are you testing on Android 8+? Then I will turn off this feature and leave it for older versions, until someone tells me what other option might be.

**Comment by stujones11** [comment]
> Are you testing on Android 8+?

Yes, I only get this problem on Android 8.

I have also noticed some issues with the full screen view.

screen

**Comment by MoNTE48** [comment]
@stujones11 remove minetest.conf to try again. My code should change the screen size faster than minetest will load and read the screen size

**Comment by stujones11** [comment]
>  remove minetest.conf to try again. 

Yes, that fixes it :+1:

**Comment by stujones11** [comment]
@MoNTE48 Sorry but the latest commit is crashing for me again with what looks like the same error as before.

**Comment by MoNTE48** [comment]
@stujones11 please, try Build -> Clean Project (in Android Studio). I tested it on my OnePlus 3 (Android 8.1)

**Comment by stujones11** [comment]
@MoNTE48 I did try that and I am sure the patch applied because it was working prior to that with your previous suggestion here https://github.com/minetest/minetest/pull/7820#issuecomment-433979453

I am building everything from scratch using a fresh clone of your repo and have removed all traces of the app from my phone. Unfortunately my internet seems to be playing up at the moment so this may take some time.

**Comment by MoNTE48** [comment]
@stujones11 https://ufile.io/2s2f1

**Comment by stujones11** [comment]
I get the same error with a clean build. I am currently downloading your apk at about 30kBps :(

**Comment by juozaspo** [comment]
@stujones11 
> > doesnt exit to phone interface when entering & leaving lock screeen.
> 
> I think this may be a pre-existing issue #7501. Please correct me if this is not what you mean,
This PR **fixes** the issue you mentioned for me, also I posted that. Posting here since this issue is mentioned in description above, so the game does not close when you lock and then unlock the phone whitch is correct, the game would crash (exit) as soon as screen were unlocked which where wrong since ~0.4.17.

**Comment by stujones11** [comment]
@juozaspo thank you for the clarification, I was not sure.

**Comment by MoNTE48** [comment]
Oh shit. It is impossible to predict the behavior of one code on one version of one operating system! That's why I love iOS.
@stujones11 https://ufile.io/5wawq THIS APK MUST WORK!

**Comment by stujones11** [comment]
> THIS APK MUST WORK!

Yes, that one works :))

**Comment by juozaspo** [comment]
@stujones11 
> I get the same error with a clean build. I am currently downloading your apk at about 30kBps :(

I built using method shown in #7797, with python and clang installed as needed (that issue was solved by that, w/o that nothing would build), on ubuntu 18.04 (bionic) vm. Builds, installs and runs fine, though on initial build there were many warnings while building dependencies such as irrlicht, no fancy stuff such as Android Studio installed. To clean everything in git tree I use `git clean -d -x -i` than I rebuild using `make debug` command in directory `build/android`.

**Comment by stujones11** [comment]
@MoNTE48 My build is working again with eb1c2ecd652cfd08c4d15d9479b4eae27c9dd3db :+1: 

> explain to Mr. nerzhul what the problem is.

Well I am not entirely sure I know what the problem is but it would appear to be pointing at the `iconv` lib.

**Comment by MoNTE48** [comment]
libc++_shared.so NOT supported Android 4. Minimal Android must be 5.0. 
Or minetest back to use gcc and prev stl, or must build two apk - for 4 and for 5+ android. Remember the work on arm64 for MC?

**Comment by stujones11** [comment]
> Or minetest back to use gcc and prev stl

This is not an option since the c++11 changes, also support for gcc is dropped completely from the latest ndk-18 which also seems to require a minimum target of api-16 to build all the deps. There may be other workarounds though, I am not sure.

**Comment by juozaspo** [comment]
>also support for gcc is dropped completely from the latest ndk-18
f611

**Comment by MoNTE48** [comment]
I was looking for workarounds some time ago. We need some kind of gcc build for the 4th android. Nerzul can include 3 apk for each update - gcc for Android 4, clang for 5+ and x86. It's not a problem. The code for such an assembly was written by you, and it is located in the repo MC

**Comment by stujones11** [comment]
> The code for such an assembly was written by you, and it is located in the repo MC

Yes, but 5.0.0-dev does not build at all with gcc since we moved to c++11 standards. It may be possible with a lot of STL hacks but that is certainly beyond my capabilities.

**Comment by MoNTE48** [comment]
And also beyond the scope of this PR. The problem is that I have not tested my code on Android 4. Although I am sure that everything is fine. I think that you will find a solution to this problem.
@SmallJoker Hi, you can add 5.0.0 tag?

**Comment by stujones11** [comment]
>  The problem is that I have not tested my code on Android 4.

Your code is probably fine since it can still be built with ndk-17 and gcc. However the 'clang' builds will likely have the same issue if what you are saying is correct.

**Comment by MoNTE48** [comment]
@rubenwardy you can test this PR?
You can compile he or install my APK from this msg: https://github.com/minetest/minetest/pull/7820#issuecomment-434018103

**Comment by paramat** [comment]
Testing appreciated @IhrFussel @codexp @benrob0329 apk linked above.

**Comment by paramat** [comment]
Tested by 3 people, and combined with MoNTE48's experience i think this counts for something, so :+1: 
However i ask that the 2nd approval comes from a core dev who understands the changes, @nerzhul maybe?

File: build/android/build.gradle [hunk 1]
```diff
@@ -20,106 +25,110 @@ def sqlite3_version = "3240000"
...
+        versionName "${System.env.VERSION_STR}.${versionCode}"
+        minSdkVersion 14
+        targetSdkVersion 28
+        applicationId "net.minetest.minetest"
+        manifestPlaceholders = [package: "net.minetest.minetest", project: project.name]
+        ndk {
+            // Specifies the ABI configurations of your native
+            // libraries Gradle should build and package with your APK.
+            // abiFilters 'armeabi-v7a', 'arm64-v8a', 'x86', 'x86_64'
+            abiFilters 'armeabi-v7a', 'x86'
```
Comment (nerzhul) [inline build/android/build.gradle]: Sorry but we must keep all architectures.

File: build/android/build.gradle [hunk 2]
```diff
@@ -1,13 +1,18 @@
...
+    repositories {
+        maven { url 'https://maven.google.com' }
+        jcenter()
+    }
+    dependencies {
+        classpath 'com.android.tools.build:gradle:3.1.3'
+    }
+}
+
+allprojects {
```
Comment (nerzhul) [inline build/android/build.gradle]: i don't see why we need this

File: build/android/build.gradle (hunk 1 above)
Comment (MoNTE48) [reply build/android/build.gradle]: If you think so, check out the google documentation or run your own tests.

**Comment by MoNTE48** [comment]
> Also, please don't reformat it in this PR, just do your additions to make the PR clear, we just need compat

It is difficult, I have to **manually restore all the lines**. I wouldn’t like to do this because you don’t like something

File: build/android/build.gradle (hunk 2 above)
Comment (MoNTE48) [reply build/android/build.gradle]: Without this, Android Studio could not successfully complete the build.

**Comment by paramat** [comment]
> please don't reformat it in this PR, just do your additions to make the PR clear

I agree, code reformatting should be a separate commit to make review easier.

**Comment by MoNTE48** [comment]
@paramat @nerzhul You can solve this question with a simple PR in this branch :)
https://github.com/MoNTE48/minetest/tree/update-java

**Comment by paramat** [comment]
Yes unfortunately it is difficult to undo, so we won't insist, but if you can that would be helpful.
However, keeping codestyle issues separate is a development rule.

**Comment by MoNTE48** [comment]
Ok, i try to restore MinetestTextEntry.java and MinetestAssetCopy.java.
WinMerge is a very powerful tool :)

**Comment by MoNTE48** [comment]
MinetestAssetCopy. 
<details>
  <summary>Click to expand</summary>







</details>

**Comment by nerzhul** [comment]
waiting for your changes, as you reformat too many files the review is a little bit difficult, it's why it's prefered

**Comment by MoNTE48** [comment]
@nerzhul I have provided you with conveniently readable screenshots. You can see what has been changed. I think you don't need screenshots of build.gradle :)?

**Comment by nerzhul** [comment]
it's not very convenient, can you just update the PR please ? you should know which change to apply without lint

**Comment by SmallJoker** [comment]
I would appreciate it if you could use tabs to indent the code lines. Rewriting with spaces does follow the guidelines and makes it hard to see the changes in the web view.

**Comment by MoNTE48** [comment]
@SmallJoker done

**Comment by stujones11** [comment]
> I would appreciate it if you could use tabs to indent the code lines.

Java code typically uses spaces for indentation (I know) and most ide's are setup to auto-format java code in this way. Android studio highlights tab indented lines with annoying warnings. Can these files not just be whitelisted?

**Comment by SmallJoker** [comment]
@MoNTE48 Thank you a lot! The diff is much more readable now.
@stujones11 I am well aware that some IDEs are pre-configured to use spaces, rather than auto-detecting the style. But in a community project like this one, the code has to be uniform. Thus, reformatting the entire Java source code only results in pseudo code changes and makes git blame harder to use. TIL Android Studio sucks.

**Comment by paramat** [comment]
I think due to the importance of this PR, we should accept either, someone can correct it later if they really want to (to whatever we eventually decide, if we ever agree).
So tabs is fine.

**Comment by paramat** [comment]
In IRC we have decided tabs is fine.

**Comment by paramat** [comment]
MoNTE48 thanks for the screenshots above but that is still somewhat difficult to review.
The functional changes should ideally be in a separate commit from codestyle and format changes, otherwise it's much more difficult to review. If you can separate them this will be much appreciated and will get this PR merged much quicker.

File: build/android/build.gradle (hunk 1 above)
Comment (paramat) [reply build/android/build.gradle]: @nerzhul happy about this?

**Comment by SmallJoker** [comment]
How mergeable is this PR? Does the latest PR HEAD work for you, @stujones11, @juozaspo ?

**Comment by stujones11** [comment]
>  Does the latest PR HEAD work for you

Works for me :+1:

**Comment by paramat** [comment]
@nerzhul how complete is your review? Do you need this split more to be more reviewable? Is the decremented loop the only remaining issue?

File: build/android/src/main/java/net.minetest.minetest/MainActivity.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+						return;
+					}
+				}
+				// permission were granted - run
+				Next();
+				break;
+		}
+	}
+
+	public void Next() {
```
Comment (rubenwardy) [inline build/android/src/main/java/net.minetest.minetest/MainActivity.java]: `next`

File: build/android/src/main/java/net.minetest.minetest/MinetestAssetCopy.java [hunk 4]
```diff
@@ -1,416 +1,373 @@
...
-	TextView m_Filename;
-	
-	copyAssetTask m_AssetCopy;
-	
-	private class copyAssetTask extends AsyncTask<String, Integer, String>
-	{
-		private long getFullSize(String filename)
-		{
+
+	private class copyAssetTask extends AsyncTask<String, Integer, String> {
```
Comment (rubenwardy) [inline build/android/src/main/java/net.minetest.minetest/MinetestAssetCopy.java]: `CopyAssetTask`

File: build/android/src/main/java/net.minetest.minetest/MinetestAssetCopy.java [hunk 5]
```diff
@@ -1,416 +1,373 @@
...
+		boolean m_copy_started = false;
```
Comment (rubenwardy) [inline build/android/src/main/java/net.minetest.minetest/MinetestAssetCopy.java]: `m_` is rarely used in Java (and is discouraged in future code in our code base)

File: build/android/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
-	<string name="preparing_media">Preparing media...</string>
-</resources>
-
+	<string name="preparing_media">Preparing media&#8230;</string>
+	<string name="not_granted">Required permission need for unpacking game data not granted. Minetest can\'t run without it</string>
```
Comment (rubenwardy) [inline build/android/src/main/res/values/strings.xml]: > Required permission wasn't granted, Minetest can't run without it!

**Comment by rubenwardy** [comment]
@paramat   please may you use the github approval feature? It's hard to see who approved otherwise, without scrolling through the whole history

**Comment by paramat** [comment]
rubenwardy, ok if it helps. But keep in mind that dev rules state an official approval is shown using a +1, LGTM, thumbs-up icon in a message, not by any other method, so i'm following rules.
(Unfortunately, a few coredevs are using 'github approval' without also using the official way to show an approval, and without using the labels, this is a problem.)

**Comment by paramat** [comment]
I'm uncertain whether i should have approved, since i do not actually understand any of the code. I would prefer this gets 2 approvals from others who do.

**Comment by juozaspo** [comment]
@SmallJoker I haven't made a build from this pr this time yet. Will make a one and do a testing a bit later today. The one made from latest master works fine.

File: build/android/src/main/AndroidManifest.xml [hunk 7]
```diff
@@ -1,34 +1,59 @@
...
+		android:glEsVersion="0x00010000"
+		android:required="true" />
+
 	<uses-permission android:name="android.permission.INTERNET" />
-	<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
 	<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
-	<application android:icon="@drawable/irr_icon"
+
+	<application
+		android:allowBackup="true"
```
[the remark this answers is not in the document]
Comment (MoNTE48) [reply build/android/src/main/AndroidManifest.xml]: No, now this not working.

File: build/android/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -1,34 +1,59 @@
...
+		android:icon="@mipmap/ic_launcher"
+		android:label="${project}"
+		android:resizeableActivity="false">
```
[the remark this answers is not in the document]
Comment (MoNTE48) [reply build/android/src/main/AndroidManifest.xml]: Minetest not support change window resolution during work (on Android)

File: build/android/src/main/AndroidManifest.xml [hunk 9]
```diff
@@ -1,34 +1,59 @@
...
+
+		<meta-data
+			android:name="android.max_aspect"
+			android:value="2.1" />
```
[the remark this answers is not in the document]
Comment (MoNTE48) [reply build/android/src/main/AndroidManifest.xml]: No. This code allow fullscreen mode for 18:9+ devices. Copied from official google play developers guide.

File: build/android/src/main/java/net.minetest.minetest/MainActivity.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+public class MainActivity extends Activity {
+
+	private final static int PERMISSIONS = 1;
+	private static final String[] REQUIRED_SDK_PERMISSIONS = new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE};
+	RelativeLayout relativeLayout;
+
+	@Override
+	public void onCreate(Bundle savedInstanceState) {
+		super.onCreate(savedInstanceState);
+		relativeLayout = new RelativeLayout(this);
```
[the remark this answers is not in the document]
Comment (MoNTE48) [reply build/android/src/main/java/net.minetest.minetest/MainActivity.java]: This code need only for background and player sees this background only 1-2 sec for all the time

File: build/android/src/main/java/net.minetest.minetest/MinetestAssetCopy.java (hunk 4 above)
Comment (MoNTE48) [reply build/android/src/main/java/net.minetest.minetest/MinetestAssetCopy.java]: This is not my code.

File: build/android/src/main/java/net.minetest.minetest/MinetestAssetCopy.java (hunk 5 above)
Comment (MoNTE48) [reply build/android/src/main/java/net.minetest.minetest/MinetestAssetCopy.java]: same

**Comment by MoNTE48** [comment]
@rubenwardy, please, check last commit.

File: build/android/src/main/java/net.minetest.minetest/MainActivity.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+					grantResults);
+		}
+	}
+
+	@Override
+	public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[],
+										   @NonNull int[] grantResults) {
+		switch (requestCode) {
+			case PERMISSIONS:
+				for (int index = permissions.length - 1; index >= 0; --index) {
```
[the remark this answers is not in the document]
Comment (paramat) [reply build/android/src/main/java/net.minetest.minetest/MainActivity.java]: Changed to increment.

File: build/android/build.gradle (hunk 1 above)
Comment (paramat) [reply build/android/build.gradle]: @nerzhul ?

**Comment by juozaspo** [comment]
@MoNTE48 Did you test your changes against latest commits in master branch? Because It looks like commits your branch is built on top are very old, canges of merged PR's such as #7834 and others are not included in your own. You should rebase against latest master then do a force push again. 
The build of minetest otherwise works fine in my phone, though I still need to fiddle around some gui scaling settings since they're broken (fixed in #7834). Game does not exit when locking and unlocking screen, no returning to the phone interface (launcher) as it should be. In-game not tested.

**Comment by MoNTE48** [comment]
@juozaspo thank's for testing.
My PR not have any conflicts, core dev can merge him without problem.

**Comment by SmallJoker** [comment]
> This is a bit of a killer:

> Because It looks like commits your branch is built on top are very old, [...]

> My PR not have any conflicts, core dev can merge him without problem.

@rubenwardy The lock screen issue is not related to this PR.

**Comment by juozaspo** [comment]
@rubenwardy It was working right in 0.4.16, then sopped working somewhere in 0.4.17. The correct behavior for the game is not to exit to launcher but to return to game when screen is unlocked, maybe pausing when screen is locked (maybe add an option for that). The issue on #7501 is that game exits to launcher instead of returning to gameplay, which is quite annoying in 0.4.17, I was used to old behavior (run game => lock the screen => maybe wait a while => unlock  the screen => be instantly returned to game w/o deliberately reopening it from launcher and reconnecting and/or restarting singleplayer)

**Comment by MoNTE48** [comment]
My pr must fixed this issue!

**Pull request merged**